src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/AuroraTopologyService.java

```diff
@@ -219,9 +219,10 @@ protected ClusterTopologyInfo queryForTopology(JdbcConnection conn) throws SQLEx
       // "Unknown table 'REPLICA_HOST_STATUS' in information_schema"
       // Ignore this kind of exceptions.
     } finally {
-      if (this.gatherPerfMetrics) {
+      if (topologyInfo != null && conn.getPropertySet() != null && conn.getPropertySet().getBooleanProperty(PropertyKey.gatherPerfMetrics.getKeyName()).getValue()) {
         long currentTimeMs = System.currentTimeMillis();
         this.queryTopologyMetrics.registerQueryExecutionTime(currentTimeMs - startTimeMs);
+        ClusterAwareMetricsContainer.linkInstances(topologyInfo.hosts, clusterId);
```

Comment:
I'm not sure it's necessary. Metrics container can be instantiated at constructor or when clusterId is set.

Reply:
Current implementation isn't testable.